src/NuGet.Core/NuGet.Protocol/Resources/VulnerabilityInfoResourceV3.cs

```diff
@@ -0,0 +1,299 @@
+                }
+
+                indexEntries = GetValidIndexEntries(indexEntries, ref exceptions);
+
+                await Task.WhenAll(tasks);
+
+                for (int i = 0; i < tasks.Length; i++)
+                {
+                    try
+                    {
+                        IReadOnlyDictionary<string, IReadOnlyList<VulnerabilityInfo>>? data = await tasks[i];
+                        if (data != null)
+                        {
+                            if (knownVulnerabilities == null)
+                            {
+                                knownVulnerabilities = new(indexEntries.Count);
+                            }
+                            knownVulnerabilities.Add(data);
+                        }
+                    }
+                    catch (Exception ex)
+                    {
+                        AddException(ref exceptions, ex);
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                AddException(ref exceptions, ex);
+            }
+
+            GetVulnerabilityInfoResult result = new(knownVulnerabilities, ToAggregateException(exceptions));
+            return result;
+
+            static void AddException(ref List<Exception>? exceptions, Exception ex)
+            {
+                if (exceptions == null)
+                {
+                    exceptions = new();
+                }
+                exceptions.Add(ex);
+            }
+
+            static AggregateException? ToAggregateException(IEnumerable<Exception>? exceptions)
+            {
+                AggregateException? aggregateException =
+                    exceptions == null
+                    ? null
+                    : new AggregateException(exceptions);
+                return aggregateException;
+            }
+        }
+
+        private IReadOnlyList<V3VulnerabilityIndexEntry> GetValidIndexEntries(IReadOnlyList<V3VulnerabilityIndexEntry> indexEntries, ref List<Exception>? exceptions)
+        {
+            List<V3VulnerabilityIndexEntry> validIndexEntries = new(indexEntries.Count);
+            HashSet<string> pageNames =
+#if NETSTANDARD
+                new(comparer: StringComparer.InvariantCultureIgnoreCase);
+#else
+                new(indexEntries.Count, StringComparer.InvariantCultureIgnoreCase);
+#endif
+
+            for (int i = 0; i < indexEntries.Count; i++)
+            {
+                V3VulnerabilityIndexEntry entry = indexEntries[i];
+
+                string name = entry.Name;
+                if (string.IsNullOrWhiteSpace(name))
+                {
+                    string message = string.Format("Vulnerability page index {0} does not contain a name (Source: '{1}')",
+                        i, _sourceRepository.PackageSource.Name);
+                    AddException(new FatalProtocolException(message), ref exceptions);
+                    continue;
+                }
```

Comment:
This method has a series of validations that follow the same pattern:
Test a predicate. If true, build a message, throw an exception, and continue.

We can reduce code duplication and make it easier to read by making a method or object for the repeated pattern.